File: Leetcode/0049-group-anagrams/0049-group-anagrams.java
class Solution {
    public List<List<String>> groupAnagrams(String[] strs) {
        Map<String, List<String>> anagramMap = new HashMap<>();

        for(String str : strs) {
            char[] charStr = str.toCharArray();
            Arrays.sort(charStr);
            String sortedStr = String.valueOf(charStr);

            if(!anagramMap.containsKey(sortedStr)) {
                anagramMap.put(sortedStr, new ArrayList<>());

            }
            
            anagramMap.get(sortedStr).add(str);
        }

        return new ArrayList<>(anagramMap.values());
    }
}

File: Leetcode/0049-group-anagrams/0049-group-anagrams.kt
class Solution {
    fun groupAnagrams(strs: Array<String>): List<List<String>> {

    val results : MutableMap<String, MutableList<String>> = mutableMapOf()

    for (str in strs) {
        val key = str.toCharArray().sorted().joinToString("")

        results.getOrPut(key) { mutableListOf() }.add(str)
    }

    return ArrayList<List<String>>(results.values)
    }
}